railties/lib/rails/generators/rails/resource_route/resource_route_generator.rb

```diff
@@ -21,7 +25,9 @@ def add_resource_route
         end
 
         # inserts the primary resource
-        write("resources :#{file_name.pluralize}", route_length + 1)
+        resource_string = "resources :#{file_name.pluralize}"
+        resource_string << ", except: [:new, :edit]" if options.api?
```

Comment:
> Regarding the routes generated by resources in API only mode, the reason for this behavior is to avoid having a routes.rb file where all the resources have the except: [:new, :edit]. Please considerar that API only apps should be applications where all the controllers inherits from AC::API and having actions like new and edit never makes sense. For this reason, it sounded odd to have to put except: [:new, :edit] in many places to achieve something that could be deduced automatically.

That is fine but we are forgetting something `api_only` is changing the behavior of all engines mounted in the application when it should not. `api_only` should only change the behavior of `resources` only in the application that configured that option. Engines should not be affected by that option.

> but I think it is not a great idea to have a way to generate AC::API controllers via generators in a regular app.

Why not? It is perfect fine to you have controllers that are used as API and controllers used with normal web views.

> Related to the previous point, a --no-api option currently exists in the generators (it's created automatically by thor). So, in an only API app AC::Base controllers can be generated, but these controllers won't work at all because an only API app does not have all the necessary middlewares to work with this kind of controllers. It's an issue that should be fixed, and it also can be fixed by removing the --api option.

:+1:

> Regarding only API apps and engines: I haven't had much time to think in all the cases, but I can affirm that mounting a normal engine in an only API app is not going to work because of the middlewares. The host app and all the engines share the same middleware stack, and this stack is defined by the api_only flag in the application, so an engine including regular controllers and using stuff like ActionView is going to fail. I don't know how we should prevent the usage of regular engines in only API apps but it has to be done somehow (at least while Rails apps have a unique shared middleware stack for the application itself and engines 😸).

Yeah that is a problem indeed. Engines will work but with some feature missing like cookies. But even with that I believe `api_only` should change the engines router. Even that engines not fully work in API only applications the router should be intact.